File: Assemblies/Protean.Tools/FTPClient.vb
Imports System
Imports System.Net
Imports System.IO
Imports System.Text
Imports System.Text.ASCIIEncoding
Imports System.Net.Sockets
Imports System.Configuration
Imports System.Resources

Public Class FTPClient

#Region "Declarations"

    Private cHost As String
    Private cUserName As String
    Private cPassword As String
    Private Const mcModuleName As String = "Eonic.Tools.FTPClient"
    Public Event OnError(ByVal sender As Object, ByVal e As Eonic.Tools.Errors.ErrorEventArgs)

#End Region

#Region "Properties"

    Public ReadOnly Property Host() As String
        Get
            Return cHost
        End Get
    End Property

    Public ReadOnly Property Username() As String
        Get
            Return cUserName
        End Get
    End Property

    Public ReadOnly Property Password() As String
        Get
            Return cPassword
        End Get
    End Property

#End Region

#Region "Public Procedures"

    Public Sub New(ByVal HostAddress As String, ByVal HostUserName As String, ByVal HostPassword As String)
        Try
            cHost = HostAddress
            cUserName = HostUserName
            cPassword = HostPassword
            Connect()
        Catch ex As Exception
            RaiseEvent OnError(Me, New Eonic.Tools.Errors.ErrorEventArgs(mcModuleName, "New", ex, ""))
        End Try
    End Sub

    Public Sub Connect()
        Try

        Catch ex As Exception
            RaiseEvent OnError(Me, New Eonic.Tools.Errors.ErrorEventArgs(mcModuleName, "Connect", ex, ""))
        End Try
    End Sub

    Public Sub Disconnect()
        Try

        Catch ex As Exception
            RaiseEvent OnError(Me, New Eonic.Tools.Errors.ErrorEventArgs(mcModuleName, "Disconnect", ex, ""))
        End Try
    End Sub

    Public Function Download(ByVal LocalFile As String, ByVal RemoteFile As String) As Boolean
        Try
            Dim outputStream As FileStream = New FileStream(LocalFile, FileMode.Create)
            Dim reqFTP As FtpWebRequest = FtpWebRequest.Create(New Uri(Replace(cHost & RemoteFile, "\", "/")))
            reqFTP.Method = WebRequestMethods.Ftp.DownloadFile
            reqFTP.UseBinary = True
            reqFTP.KeepAlive = True
            reqFTP.Credentials = New NetworkCredential(cUserName, cPassword)
            Dim response As FtpWebResponse = reqFTP.GetResponse()
            Dim ftpstream As Stream = response.GetResponseStream()
            Dim cl As Long = response.ContentLength
            Dim buffersize As Integer = 2048
            Dim readcount As Integer
            Dim buffer(buffersize) As Byte
            readcount = ftpstream.Read(buffer, 0, buffersize)
            Do While (readcount > 0)
                outputStream.Write(buffer, 0, readcount)
                readcount = ftpstream.Read(buffer, 0, buffersize)
            Loop
            ftpstream.Close()
            outputStream.Close()
            response.Close()
            Return True
        Catch ex As Exception
            RaiseEvent OnError(Me, New Eonic.Tools.Errors.ErrorEventArgs(mcModuleName, "Download", ex, ""))
            Return False
        End Try
    End Function

    Public Function UploadFile(ByVal LocalFile As String, ByVal RemoteFile As String) As Boolean
        Try
            Dim reqFTP As FtpWebRequest = FtpWebRequest.Create(New Uri(Replace(cHost & RemoteFile, "\", "/")))
            reqFTP.Method = WebRequestMethods.Ftp.UploadFile
            reqFTP.UseBinary = True
            reqFTP.Proxy = Nothing
            reqFTP.KeepAlive = True
            reqFTP.Credentials = New NetworkCredential(cUserName, cPassword)
            Dim sourceStream As New StreamReader(LocalFile)
            Dim fileContents() As Byte = System.Text.Encoding.UTF8.GetBytes(sourceStream.ReadToEnd())
            sourceStream.Close()
            reqFTP.ContentLength = fileContents.Length
            Dim requestStream As Stream = reqFTP.GetRequestStream()
            requestStream.Write(fileContents, 0, fileContents.Length)
            requestStream.Close()
            Return True
        Catch ex As Exception
            RaiseEvent OnError(Me, New Eonic.Tools.Errors.ErrorEventArgs(mcModuleName, "Upload", ex, ""))
            Return False
        End Try
    End Function

    Public Function UploadText(ByVal FileText As String, ByVal RemotePath As String, ByVal FileName As String) As Boolean
        Try
            Dim ftpRequest As FtpWebRequest = FtpWebRequest.Create(New Uri(cHost & RemotePath & FileName))
            ftpRequest.Method = WebRequestMethods.Ftp.UploadFile
            ftpRequest.Proxy = Nothing
            ftpRequest.UseBinary = True
            ftpRequest.Credentials = New NetworkCredential(cUserName, cPassword)
            Dim fileContents As Byte() = System.Text.Encoding.UTF8.GetBytes(FileText)
            Dim writer As Stream = ftpRequest.GetRequestStream()
            Debug.WriteLine("[" & Now.ToLongTimeString & "] Sending File")
            writer.Write(fileContents, 0, fileContents.Length)
            Return True
        Catch ex As Exception
            RaiseEvent OnError(Me, New Eonic.Tools.Errors.ErrorEventArgs(mcModuleName, "Upload", ex, ""))
            Return False
        End Try
    End Function

#End Region

End Class

File: Assemblies/Protean.Tools/FTPClient.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Text;

namespace Protean.Tools
{

    public class FTPClient
    {

        #region Declarations

        private string cHost;
        private string cUserName;
        private string cPassword;
        private const string mcModuleName = "Protean.Tools.FTPClient";
        public event OnErrorEventHandler OnError;

        public delegate void OnErrorEventHandler(object sender, Protean.Tools.Errors.ErrorEventArgs e);

        #endregion

        #region Properties

        public string Host
        {
            get
            {
                return cHost;
            }
        }

        public string Username
        {
            get
            {
                return cUserName;
            }
        }

        public string Password
        {
            get
            {
                return cPassword;
            }
        }

        #endregion

        #region Public Procedures

        public FTPClient(string HostAddress, string HostUserName, string HostPassword)
        {
            try
            {
                cHost = HostAddress;
                cUserName = HostUserName;
                cPassword = HostPassword;
                Connect();
            }
            catch (Exception ex)
            {
                OnError?.Invoke(this, new Protean.Tools.Errors.ErrorEventArgs(mcModuleName, "New", ex, ""));
            }
        }

        public void Connect()
        {
            try
            {
            }

            catch (Exception ex)
            {
                OnError?.Invoke(this, new Protean.Tools.Errors.ErrorEventArgs(mcModuleName, "Connect", ex, ""));
            }
        }

        public void Disconnect()
        {
            try
            {
            }

            catch (Exception ex)
            {
                OnError?.Invoke(this, new Protean.Tools.Errors.ErrorEventArgs(mcModuleName, "Disconnect", ex, ""));
            }
        }

        public bool Download(string LocalFile, string RemoteFile)
        {
            try
            {
                var outputStream = new FileStream(LocalFile, FileMode.Create);
                FtpWebRequest reqFTP = (FtpWebRequest)WebRequest.Create(new Uri(cHost + RemoteFile.Replace(@"\", "/")));
                reqFTP.Method = WebRequestMethods.Ftp.DownloadFile;
                reqFTP.UseBinary = true;
                reqFTP.KeepAlive = true;
                reqFTP.Credentials = new NetworkCredential(cUserName, cPassword);
                FtpWebResponse response = (FtpWebResponse)reqFTP.GetResponse();
                Stream ftpstream = response.GetResponseStream();
                long cl = response.ContentLength;
                int buffersize = 2048;
                int readcount;
                var buffer = new byte[buffersize + 1];
                readcount = ftpstream.Read(buffer, 0, buffersize);
                while (readcount > 0)
                {
                    outputStream.Write(buffer, 0, readcount);
                    readcount = ftpstream.Read(buffer, 0, buffersize);
                }
                ftpstream.Close();
                outputStream.Close();
                response.Close();
                return true;
            }
            catch (Exception ex)
            {
                OnError?.Invoke(this, new Protean.Tools.Errors.ErrorEventArgs(mcModuleName, "Download", ex, ""));
                return false;
            }
        }

        public bool UploadFile(string LocalFile, string RemoteFile)
        {
            try
            {
                FtpWebRequest reqFTP = (FtpWebRequest)WebRequest.Create(new Uri(cHost + RemoteFile.Replace(@"\", "/")));
                reqFTP.Method = WebRequestMethods.Ftp.UploadFile;
                reqFTP.UseBinary = true;
                reqFTP.Proxy = null;
                reqFTP.KeepAlive = true;
                reqFTP.Credentials = new NetworkCredential(cUserName, cPassword);
                var sourceStream = new StreamReader(LocalFile);
                byte[] fileContents = Encoding.UTF8.GetBytes(sourceStream.ReadToEnd());
                sourceStream.Close();
                reqFTP.ContentLength = fileContents.Length;
                Stream requestStream = reqFTP.GetRequestStream();
                requestStream.Write(fileContents, 0, fileContents.Length);
                requestStream.Close();
                return true;
            }
            catch (Exception ex)
            {
                OnError?.Invoke(this, new Protean.Tools.Errors.ErrorEventArgs(mcModuleName, "Upload", ex, ""));
                return false;
            }
        }

        public bool UploadText(string FileText, string RemotePath, string FileName)
        {
            try
            {
                FtpWebRequest ftpRequest = (FtpWebRequest)WebRequest.Create(new Uri(cHost + RemotePath + FileName));
                ftpRequest.Method = WebRequestMethods.Ftp.UploadFile;
                ftpRequest.Proxy = null;
                ftpRequest.UseBinary = true;
                ftpRequest.Credentials = new NetworkCredential(cUserName, cPassword);
                byte[] fileContents = Encoding.UTF8.GetBytes(FileText);
                Stream writer = ftpRequest.GetRequestStream();
                Debug.WriteLine("[" + DateTime.Now.ToLongTimeString() + "] Sending File");
                writer.Write(fileContents, 0, fileContents.Length);
                return true;
            }
            catch (Exception ex)
            {
                OnError?.Invoke(this, new Protean.Tools.Errors.ErrorEventArgs(mcModuleName, "Upload", ex, ""));
                return false;
            }
        }

        #endregion

    }
}
